hal/src/nRF52840/exflash_hal.cpp

```
@@ -0,0 +1,739 @@
+            data_to_process_this_operation = HAL_QSPI_FLASH_OTP_SECTOR_SIZE_GD25 - register_offset;
+        }
+
+        ret = operation(otp_page, register_offset, &data_buf[data_processed], data_to_process_this_operation);
+
+        // Increment/decrement counter variables, increment to the next OTP register
+        data_to_process -= data_to_process_this_operation;
+        data_processed += data_to_process_this_operation;
+        otp_page++;
+    }
+
+    return ret;
+}
+
+int erase_common(uintptr_t start_addr, size_t num_blocks, nrf_qspi_erase_len_t len) {
+    FlashLock lk;
+
+    int err_code = NRF_SUCCESS;
+
+    const size_t block_length = len == NRF_QSPI_ERASE_LEN_4KB ? 4096 : 64 * 1024;
+
+    /* Round down to the nearest block */
+    start_addr = ((start_addr / block_length) * block_length);
+
+    for (size_t i = 0; i < num_blocks; i++) {
+        err_code = nrfx_qspi_erase(len, start_addr);
+        CHECK_NRF_RETURN(err_code, nrf_system_error(err_code));
+
+        exflash_qspi_wait_completion();
+
+        start_addr += block_length;
+    }
+
+    // LOG_DEBUG(TRACE, "Erased %lu %lukB blocks starting from %" PRIxPTR,
+    //           num_blocks, block_length / 1024, start_addr);
+
+    return 0;
+}
+
+} // Anonymous
+
+int hal_exflash_init(void) {
+    // PATCH: Initialize CS pin, external memory discharge
+    nrf_gpio_cfg_output(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_pin_set(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_cfg_input(QSPI_FLASH_IO1_PIN, NRF_GPIO_PIN_PULLDOWN);
+
+    FlashLock lk;
+
+    hal_qspi_flash_type_t all_flashes[] = {
+        HAL_QSPI_FLASH_TYPE_MX25L3233F,
+        HAL_QSPI_FLASH_TYPE_MX25R6435F,
+        HAL_QSPI_FLASH_TYPE_GD25WQ64E
+    };
+
+    // Use single data line SPI to read ID
+    nrfx_qspi_config_t default_config = NRF_QSPI_QUICK_CFG(NRF_QSPI_WRITEOC_PP, NRF_QSPI_READOC_FASTREAD, NRF_QSPI_FREQ_32MDIV2);
+    int ret = nrfx_qspi_init(&default_config, nullptr, nullptr);
+    if (ret == NRFX_ERROR_TIMEOUT) {
```

When the GD25 is in deep sleep, `nrfx_qspi_init` does not return an error code. The first indication of the chip being in deep sleep is when the `HAL_EXFLASH_COMMAND_READID` command below fails. 
I believe the `NRFX_ERROR_TIMEOUT` error condition only happens when the chip really is doing a program/erase operation, so if this happens do the `SUSPEND_PGMERS` operation for all known chips

---

👍 @scott-brust I reproduced this behavior, it is different from the document, the document shows the status register can be read at any time. Also agree that we don't need to return an error since we have `SUSPEND_PGMERS` + `RESET`.

Should we handle other error conditions of `nrfx_qspi_init()`?

---

Yup, it does put things into the right state and it's exactly how you imagine: we've seen resets (either with a button or during OTA or especially brownouts) triggering a condition where the flash is busy on next boot and we have to explicitly abort the operation.